nav2_core/include/nav2_core/global_planner.hpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019, Samsung Research America
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef NAV2_CORE_GLOBAL_PLANNER_H_
+#define NAV2_CORE_GLOBAL_PLANNER_H_
+
+#include <string>
+#include "rclcpp/rclcpp.h"
+#include "nav2_costmap_2d/nav2_costmap_2d.hpp"
+#include "tf2_ros/Buffer.h"
+#include "nav2_msgs/msg/path.h"
+#include "geometry_msgs/msg/pose_stamped.h"
+
+namespace nav2_core
+{
+
+/**
+ * @class GlobalPlanner
+ * @brief Abstract interface for global planners to adhere to with pluginlib
+ */
+class GlobalPlanner
+{
+public:
+  /**
+   * @brief Virtual destructor
+   */
+  virtual ~GlobalPlanner() {}
+
+  /**
+   * @param  parent pointer to user's node
+   * @param  name The name of this planner
+   * @param  tf A pointer to a TF buffer
+   * @param  costmap_ros A pointer to the costmap
+   */
+  virtual void configure(const rclcpp::Node * parent,
```

Comment:
What if we need to pass an `rclcpp::LifecycleNode`?
What overloads to `configure` should we enforce? `rclcpp::Node` and `rclcpp_lifecycle::LifecycleNode`, `node_interfaces` only, or all three?

Reply:
what do you mean overloads? Why do there need to be any overloads. If we have a plugin factory we just call the configure() methods in our lifecycle `on_configure()` for the planning server. We could call it `on_configure()` too, but then may as well just derive from  `LifecycleNodeInterface` are we're talking about above

Reply:
Doesn't lifecycle node have in it a rclcpp::Node to hand off? that seems to be the most portable

Reply:
Also, what will the Planner use the node for?

Reply:
Parameters, topics, services, etc anything any other plugin might want.

Reply:
If we have a planning server, really its goal is to implement the action server, load the plugins, and pass information / retain state. The NavFn implemented plugin to that would need to potentially transform poses, or another planner may want to take in data from other sources like human tracks (needs topic) or transform poses from base link to map (TF) to then compute the time parameterized path.

Reply:
> Doesn't lifecycle node have in it a rclcpp::Node to hand off? that seems to be the most portable

There wasn't one in Dashing, that is the reason for my question

Reply:
Ah got it, then lets do it.

Reply:
So my action from this is to replace Node with LifeCycleNode ptr as the input

Reply:
> So my action from this is to replace Node with LifecycleNode ptr as the input
 
That would solve our current need. -- However, I'm not sure if in the future we might want to pass a regular node.

Now, if we're passing an `rclcpp::LifecycleNode`, is this another reason for matching the function calls for state transitions in `nav2_core::GlobalPlanner`? -- Which I believe you've already added here.

Reply:
> Parameters, topics, services, etc anything any other plugin might want.

> If we have a planning server, really its goal is to implement the action server, load the plugins, and pass information / retain state. The NavFn implemented plugin to that would need to potentially transform poses, or another planner may want to take in data from other sources like human tracks (needs topic) or transform poses from base link to map (TF) to then compute the time parameterized path.

The reason for asking about the need for a node was to explore if we could abstract a bit more this interface. How would we design this interface if not using ROS objects? :thinking: We would probably rely more on concepts to express what the planner might need.

Just something I'm interested in hearing your thoughts.

Reply:
> Now, if we're passing an rclcpp::LifecycleNode, is this another reason for matching the function calls for state transitions in nav2_core::GlobalPlanner? -- Which I believe you've already added here.

I don't understand what you're getting at. We need a node pointer in order to access things in the ros side of the house, which a planner or any thing of substance will require. The lifecycle node pointer is not to implement a lifecycle node of the plugin, its only to create subs/pubs/servers/params/etc. It should not derive from a node because we dont want each plugin to be a node.

Reply:
ROS1 would allow anyone to casually make a NodeHandle, we cannot do that in ROS2 so we we need to provide this node pointer for the plugin

Reply:
I think you might consider either overloading to use different types of nodes, using a template, or take in just the node interfaces (node_topic, node_base, node_logging, etc.)

Reply:
Interesting thought. As long as they implement the `create_subscriber` `declare_parameter` `etc` methods needed, I don't see why it should hurt. 

Then the server would have to provide the template, which in our case would always be a lifecycle node. Is there a situation we'd want to back out of lifecycle for the action servers in favor of another base type?

Reply:
The API has been updated for LifecycleNode

Reply:
> I don't understand what you're getting at. We need a node pointer in order to access things in the ros side of the house, which a planner or any thing of substance will require. The lifecycle node pointer is not to implement a lifecycle node of the plugin, its only to create subs/pubs/servers/params/etc. It should not derive from a node because we dont want each plugin to be a node.

Sorry, I've should've provided more info. So even if the plugin is not a Lifecycle node itself, the subs/pubs/servers/params created from a LifecycleNode might need to transitioned, that is currently the case for the [LifecyclePublisher](https://github.com/ros2/rclcpp/blob/1fff8cbac14319b037ca1af8a9a0e19e9b2247f7/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_publisher.hpp#L36), so at least we need the `on_activate` and `on_deactivate` on the plugin:

```cpp
void GlobalPlanner::on_activate(){
  lifecycle_pub_->on_activate();
 ...
}

void GlobalPlanner::on_deactivate(){
  lifecycle_pub_->on_deactivate();
  ...
}
...
```

Now, I'm not sure what the plans are for lifecycle versions of subs/pubs/servers/params, and if all might implement all lifecycle transitions. Just something to keep in mind.